vmm/src/serial_manager.rs

```diff
@@ -0,0 +1,245 @@
+    serial: Arc<Mutex<Pl011>>,
+    epoll_file: File,
+    in_file: File,
+    kill_evt: EventFd,
+    handle: Option<thread::JoinHandle<()>>,
+}
+
+impl SerialManager {
+    pub fn new(
+        #[cfg(target_arch = "x86_64")] serial: Arc<Mutex<Serial>>,
+        #[cfg(target_arch = "aarch64")] serial: Arc<Mutex<Pl011>>,
+        pty_pair: Option<Arc<Mutex<PtyPair>>>,
+        mode: ConsoleOutputMode,
+    ) -> Result<Option<Self>> {
+        let in_file = match mode {
+            ConsoleOutputMode::Pty => {
+                if let Some(pty_pair) = pty_pair {
+                    pty_pair
+                        .lock()
+                        .unwrap()
+                        .main
+                        .try_clone()
+                        .map_err(Error::FileClone)?
+                } else {
+                    return Ok(None);
+                }
+            }
+            ConsoleOutputMode::Tty => {
+                // If running on an interactive TTY then accept input
+                if unsafe { libc::isatty(libc::STDIN_FILENO) == 1 } {
+                    let stdin_clone = unsafe { File::from_raw_fd(libc::dup(libc::STDIN_FILENO)) };
+                    let ret = unsafe {
+                        let mut flags = libc::fcntl(stdin_clone.as_raw_fd(), libc::F_GETFL);
+                        flags |= libc::O_NONBLOCK;
+                        libc::fcntl(stdin_clone.as_raw_fd(), libc::F_SETFL, flags)
+                    };
+
+                    if ret < 0 {
+                        return Err(Error::SetNonBlocking(std::io::Error::last_os_error()));
+                    }
+
+                    stdin_clone
+                } else {
+                    return Ok(None);
+                }
+            }
+            _ => return Ok(None),
+        };
+
+        let epoll_fd = epoll::create(true).map_err(Error::Epoll)?;
+        let kill_evt = EventFd::new(EFD_NONBLOCK).map_err(Error::EventFd)?;
+
+        epoll::ctl(
+            epoll_fd,
+            epoll::ControlOptions::EPOLL_CTL_ADD,
+            kill_evt.as_raw_fd(),
+            epoll::Event::new(epoll::Events::EPOLLIN, EpollDispatch::Kill as u64),
+        )
+        .map_err(Error::Epoll)?;
+
+        epoll::ctl(
+            epoll_fd,
+            epoll::ControlOptions::EPOLL_CTL_ADD,
+            in_file.as_raw_fd(),
+            epoll::Event::new(epoll::Events::EPOLLIN, EpollDispatch::File as u64),
+        )
+        .map_err(Error::Epoll)?;
+
+        // Use 'File' to enforce closing on 'epoll_fd'
+        let epoll_file = unsafe { File::from_raw_fd(epoll_fd) };
+
+        Ok(Some(SerialManager {
+            serial,
+            epoll_file,
+            in_file,
+            kill_evt,
+            handle: None,
+        }))
+    }
+
+    pub fn start_thread(&mut self, exit_evt: EventFd) -> Result<()> {
+        // Don't allow this to be run if the handle exists
+        if self.handle.is_some() {
+            return Ok(());
+        }
```

Comment:
Nice, you're handling the case where this function would be called multiple times. Might be worth adding a warning message here because we don't expect this to happen.